### PR TITLE
feat: No-mint version of `withdraw-nft-asset`

### PR DESCRIPTION
**Pull request by gregorycoppola**
### Description

This adds a no-mint version of the NFT part of the "withdraw" functions from the hyper-chain.

I'm leaving FT for after the review because it will be analogous.

### Applicable issues
- towards #106

### Additional info (benefits, drawbacks, caveats)

This PR involves a lot of copy-and-pasting code.

This seems like the best we can do at this point. If we had the ability to wrap a trait reference in an optional in clarity, we could save a lot of code copying, but it doesn't work right now https://discord.com/channels/621759717756370964/713087894260023377/989921849318191104

### Checklist
- [x] Test coverage for new or modified code paths

### CODE REVIEW

File: core-contracts/contracts/helper/simple-nft-no-mint.clar [hunk 1]
```diff
@@ -0,0 +1,47 @@
+;; This is like `simple-nft.clar`, but does not support minting.
```
Comment (pavitthrap) [inline core-contracts/contracts/helper/simple-nft-no-mint.clar:1]: this seems identical to `simple-nft.clar` to me?

File: core-contracts/tests/hyperchains/hyperchains_test.ts [hunk 2]
```diff
@@ -1052,4 +1052,172 @@ Clarinet.test({
...
+                    types.buff(nft_leaf_hash),
+                    types.list([types.tuple({
+                        "hash": types.buff(nft_sib_hash),
+                        "is-left-side": types.bool(true)
+                    })])
+                ],
+                miner.address),
+        ]);
+
+        block.receipts[0].result
```
Comment (pavitthrap) [inline core-contracts/tests/hyperchains/hyperchains_test.ts:1119]: Add comment with name of expected error here

**Review by gregorycoppola — COMMENTED** [review]
I added a new test for the case where one user deposits an NFT, but it gets withdrawn to another user, in the no-mint case.

File: core-contracts/contracts/hyperchains.clar [hunk 3]
```diff
@@ -410,6 +413,62 @@
...
+;; Like `inner-transfer-or-mint-nft-asset but without allowing or requiring a mint function. In order to withdraw, the user must
+;; have the appropriate balance.
+(define-private (inner-transfer-without-mint-nft-asset (id uint) (recipient principal) (nft-contract <nft-trait>))
+    (let (
+            (call-result (contract-call? nft-contract get-owner id))
+            (nft-owner (unwrap! call-result (err ERR_CONTRACT_CALL_FAILED)))
+            (contract-owns-nft (is-eq nft-owner (some CONTRACT_ADDRESS)))
+        )
+
+        (print { contract-owns-nft: contract-owns-nft })
```
Comment (pavitthrap) [inline core-contracts/contracts/hyperchains.clar]: why this print?

File: core-contracts/contracts/hyperchains.clar [hunk 4]
```diff
@@ -410,6 +413,62 @@
...
+)
+
+;; Like `inner-withdraw-nft-asset` but without allowing or requiring a mint function. In order to withdraw, the user must
+;; have the appropriate balance.
+(define-public (inner-withdraw-nft-asset-no-mint (id uint) (recipient principal) (nft-contract <nft-trait>) (withdrawal-root (buff 32)) (withdrawal-leaf-hash (buff 32)) (sibling-hashes (list 50 (tuple (hash (buff 32)) (is-left-side bool) ) )))
+    (let (
+            (hashes-are-valid (check-withdrawal-hashes withdrawal-root withdrawal-leaf-hash sibling-hashes))
+         )
+
+        (print { checkpoint: "inner-withdraw-nft-asset", hashes-are-valid: hashes-are-valid })
```
Comment (pavitthrap) [inline core-contracts/contracts/hyperchains.clar]: could you remove the prints?

File: core-contracts/contracts/hyperchains.clar [hunk 5]
```diff
@@ -410,6 +413,62 @@
     )
 )
 
+;; Like `inner-transfer-or-mint-nft-asset but without allowing or requiring a mint function. In order to withdraw, the user must
+;; have the appropriate balance.
+(define-private (inner-transfer-without-mint-nft-asset (id uint) (recipient principal) (nft-contract <nft-trait>))
```
[the remark this answers is not in the document]
Comment (pavitthrap) [reply core-contracts/contracts/hyperchains.clar]: I would just add the assertion `(asserts! contract-owns-nft (err ERR_NFT_NOT_OWNED_BY_CONTRACT))` to `inner-transfer-nft-asset`

File: core-contracts/contracts/hyperchains.clar [hunk 6]
```diff
@@ -410,6 +413,62 @@
...
+        (print { contract-owns-nft: contract-owns-nft })
+
+        (asserts! contract-owns-nft (err ERR_NFT_NOT_OWNED_BY_CONTRACT))
+        (inner-transfer-nft-asset id CONTRACT_ADDRESS recipient nft-contract)
+    )
+)
+
+;; Like `inner-withdraw-nft-asset` but without allowing or requiring a mint function. In order to withdraw, the user must
+;; have the appropriate balance.
+(define-public (inner-withdraw-nft-asset-no-mint (id uint) (recipient principal) (nft-contract <nft-trait>) (withdrawal-root (buff 32)) (withdrawal-leaf-hash (buff 32)) (sibling-hashes (list 50 (tuple (hash (buff 32)) (is-left-side bool) ) )))
```
[the remark this answers is not in the document]
Comment (pavitthrap) [reply core-contracts/contracts/hyperchains.clar]: hmm I meant near `withdraw-nft-asset` at line 202

File: core-contracts/tests/hyperchains/hyperchains_test.ts [hunk 7]
```diff
@@ -1052,4 +1052,172 @@ Clarinet.test({
...
+        ]);
+        assertEquals(block.height, 3);
+        block.receipts[0].result
+            .expectOk()
+            .expectBuff(new Uint8Array([0, 1, 1, 1, 1]));
+
+        let nft_sib_hash = new Uint8Array([33, 202, 115, 15, 237, 187, 156, 88, 59, 212, 42, 195, 30, 149, 130, 0, 37, 203, 93, 165, 189, 33, 107, 213, 116, 211, 170, 0, 89, 231, 154, 3]);
+        let nft_leaf_hash = new Uint8Array([38, 72, 158, 13, 57, 120, 9, 95, 13, 62, 11, 118, 71, 237, 60, 173, 121, 221, 127, 38, 163, 75, 203, 191, 227, 4, 195, 17, 239, 76, 42, 55]);
+
+        // Miner should be able to withdraw NFT asset
```
[the remark this answers is not in the document]
Comment (pavitthrap) [reply core-contracts/tests/hyperchains/hyperchains_test.ts]: "// Miner should *not* be able to withdraw NFT asset"

File: core-contracts/contracts/hyperchains.clar (hunk 3 above)
Comment (gregorycoppola) [reply core-contracts/contracts/hyperchains.clar]: forgot to push latest :/

File: core-contracts/contracts/hyperchains.clar (hunk 4 above)
Comment (gregorycoppola) [reply core-contracts/contracts/hyperchains.clar]: forgot to push latest :/

File: core-contracts/contracts/helper/simple-nft-no-mint.clar (hunk 1 above)
Comment (gregorycoppola) [reply core-contracts/contracts/helper/simple-nft-no-mint.clar]: marking resolved

File: core-contracts/contracts/hyperchains.clar (hunk 6 above)
[the remark this answers is not in the document]
Comment (gregorycoppola) [reply core-contracts/contracts/hyperchains.clar]: I think they are next to now.. unless there was a particular order.

File: core-contracts/tests/hyperchains/hyperchains_test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (gregorycoppola) [reply core-contracts/tests/hyperchains/hyperchains_test.ts]: comments updated

**Comment by kantai** [comment]
These changes look fine to me, but this PR needs to be updated from the main branch.

**Comment by saralab** [comment]
@gregorycoppola Update from master and merge today

**Comment by gregorycoppola** [comment]
Merged `master` back in to this branch.

**Review by kantai — APPROVED** [review]
LGTM, just make sure all tests pass before merging

**Pull request merged**